File: DS/Trie/trie.cpp
#include <string>
using namespace std;

class TrieNode {
public:
    TrieNode *child[26];
    bool word;
    TrieNode(char ch) {
        for (int i = 0; i < 26; ++i)
            child[i] = nullptr;
        word = false;
    }
};
class Trie {
public:
    Trie() {
        root = new TrieNode(' ');
    }
    ~Trie() {
        delete root;
    }
    
    void insert(string word) {
        TrieNode *curr = root;
        for (char ch: word) {
            if (curr->child[ch - 'a'] == nullptr)
                curr->child[ch - 'a'] = new TrieNode(ch - 'a');
            curr = curr->child[ch - 'a'];
        }
        curr->word = true;
    }
    
    bool search(string word) {
        TrieNode *curr = root;
        for (char ch: word) {
            if (curr->child[ch - 'a'] == nullptr)
                return false;
            curr = curr->child[ch - 'a'];
        }
        return curr->word;
    }

    // return true if there is a word that has a prefix
    bool startsWith(string prefix) {
        TrieNode *curr = root;
        for (char ch: prefix) {
            if (curr->child[ch - 'a'] == nullptr)
                return false;
            curr = curr->child[ch - 'a'];
        }
        return true;
    }
private:
    TrieNode *root;
};

File: DS/Trie/trie.py

class Trie:
    def __init__(self):
        self.word = False
        self.dic = {}

    def insert(self, word: str) -> None:
        tree = self
        for w in word:
            if w not in tree.dic:
                tree.dic[w] = Trie()
            tree = tree.dic[w]
        tree.word = True

    def search(self, word: str) -> bool:
        tree = self
        for w in word:
            if w not in tree.dic:
                return False
            tree = tree.dic[w]
        return tree.word
    
    def startsWith(self, prefix: str) -> bool:
        tree = self
        for w in prefix:
            if w not in tree.dic:
                return False
            tree = tree.dic[w]
        return True